Core/SubterfugeCore/Core/Timing/TimeMachine.cs

```diff
@@ -1,166 +1,258 @@
+				Console.WriteLine(futureToPast.GetType().ToString().Substring(futureToPast.GetType().ToString().LastIndexOf('.') + 1));
+				futureToPast.ForwardAction(this, _gameState);
+				this._pastEventQueue.Enqueue(futureToPast);
+			}
+		}
+
+		/// <summary>
+		/// Reverts the game state to the previous tick. Vision events made in the current (or future) ticks are not altered.
+		/// </summary>
+		public void PreviousTick()
+		{
+			while (_pastEventQueue.Count > 0 && _pastEventQueue.Peek().GetOccursAt() >= this.GetCurrentTick())
+			{
+				// Move commands from the past to the future
+				GameEvent pastToFuture = _pastEventQueue.Dequeue();
+				pastToFuture.BackwardAction(this, _gameState);
+				this._futureEventQueue.Enqueue(pastToFuture);
+			}
+			this._gameState.CurrentTick = this._gameState.CurrentTick.Rewind(1);
+		}
+
+		/// <summary>
+		/// Jumps to a specific GameTick. Does not set vision events.
+		/// </summary>
+		/// <param name="tick">The GameTick to jump to</param>
+		public void GoTo(GameTick tick)
+		{
+			if (tick > this._gameState.CurrentTick)
+			{
+				Advance(tick - this._gameState.CurrentTick);
+			}
+			else if (tick < this._gameState.CurrentTick)
+			{
+				Rewind(this._gameState.CurrentTick - tick);
+			}
+		}
+
+		/// <summary>
+		/// Gets the GameTick that the time machine is currently representing.
+		/// </summary>
+		/// <returns>The GameTick that the timeMachine is showing</returns>
+		public GameTick GetCurrentTick()
+		{
+			return _gameState.CurrentTick;
+		}
+
+
+		/// <summary>
+		/// Jumps to a specific GameEvent
+		/// </summary>
+		/// <param name="eventOfInterest">The GameEvent to jump to</param>
+		public void GoTo(GameEvent eventOfInterest)
+		{
+			this.GoTo(eventOfInterest.GetOccursAt());
+		}
+
+		/// <summary>
+		/// For debugging. Advances the timeMachine by a specified number of ticks.
+		/// </summary>
+		/// <param name="ticks">The number of ticks to advance</param>
+		public void Advance(int ticks)
+		{
+			for (int i = 0; i < ticks; i++)
+			{
+				NextTick();
+			}
+		}
+
+		/// <summary>
+		/// For debugging. Rewinds the timeMachine by a specified number of ticks.
+		/// </summary>
+		/// <param name="ticks">The number of ticks to rewind</param>
+		public void Rewind(int ticks)
+		{
+			for (int i = 0; i < ticks; i++)
+			{
+				PreviousTick();
+			}
+		}
+
+		/// <summary>
+		/// Gets a list of all game events that occur on a given tick.
+		/// </summary>
+		/// <param name="gameTick">The game tick to get the events of.</param>
+		/// <returns>A list of all game evnts that occur on a given tick. Returns an empty list if there are none.</returns>
+		public List<GameEvent> GetEventsOnTick(GameTick gameTick)
+		{
+			if (gameTick > this.GetCurrentTick())
+			{
+				List<GameEvent> futureEventList = this._futureEventQueue.GetQueue();
+				if (futureEventList.Count == 0)
+				{
+					return new List<GameEvent>();
+				}
+				// Binary search 
```

Comment:
What is the binary search for here?

Could you not just `Advance(gameTick)` peek the queue, and then `rewind`?